File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let answer = 0;
    let visited = Array(dungeons.length).fill(false);

    function DFS(k, count) {
        for (let i = 0; i < dungeons.length; i++) {
            if (k >= dungeons[i][0] && !visited[i]) {
                visited[i] = true;
                DFS(k - dungeons[i][1], count + 1);
                visited[i] = false; 
            }
        }
        answer = Math.max(answer, count);
    }

    DFS(k, 0);

    return answer;
}

// 최소 필요 피로도 : 각 던전마다 탐험을 시작하기 위해 가지고 있어야됨
// 소모 피로도 : 던전 탐험을 마쳤을 때 소모됨

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    def dfs(k, dungeons, visited, cnt, maxcnt):
        maxcnt[0] = max(maxcnt[0], cnt)
        for i in range(len(dungeons)):
            if visited[i] == 0 and k >= dungeons[i][0]:
                visited[i] = 1
                dfs(k - dungeons[i][1], dungeons, visited, cnt + 1, maxcnt)
                visited[i] = 0
        return maxcnt
    
    maxcnt = [0]
    visited = [0 for _ in range(len(dungeons))]
    return dfs(k, dungeons, visited, 0, maxcnt)[0]